File: sdks/typescript/src/pollRender.ts
import { RendersApi } from '../generated/src/apis/RendersApi';
import type { Render } from '../generated/src/models/Render';

const TERMINAL_STATUSES = ['succeeded', 'failed', 'cancelled'] as const;

export interface PollRenderOptions {
  /** Total wall-clock time to wait before giving up. Default 60s. */
  maxWaitMs?: number;
  /** Initial interval. Subsequent polls back off until `maxIntervalMs`. Default 500ms. */
  initialIntervalMs?: number;
  /** Hard ceiling on the polling interval. Default 5s. */
  maxIntervalMs?: number;
  /** Multiplicative backoff factor between polls. Default 1.6. */
  backoffFactor?: number;
  /** Abort the poll loop early. */
  signal?: AbortSignal;
}

/**
 * Poll a Docgen render until it reaches a terminal state.
 *
 * The generated SDK exposes `getRender`, but spamming it every 50ms is
 * wasteful and gets you rate-limited. This helper:
 *
 *   - backs off exponentially (500ms → 800ms → 1.28s → … → 5s cap)
 *   - respects an overall `maxWaitMs` budget (default 60s)
 *   - cooperates with an `AbortSignal` so you can wire it to UI cancel
 *
 * Throws `PollRenderTimeout` if `maxWaitMs` elapses before terminal.
 * Throws `PollRenderAborted` if the signal fires.
 */
export async function pollRender(
  api: RendersApi,
  renderId: string,
  options: PollRenderOptions = {},
): Promise<Render> {
  const {
    maxWaitMs = 60_000,
    initialIntervalMs = 500,
    maxIntervalMs = 5_000,
    backoffFactor = 1.6,
    signal,
  } = options;

  const started = Date.now();
  let interval = initialIntervalMs;

  // eslint-disable-next-line no-constant-condition
  while (true) {
    if (signal?.aborted) {
      throw new PollRenderAborted(renderId);
    }

    const render = await api.getRender({ renderId });

    if (isTerminal(render.status)) {
      return render;
    }

    if (Date.now() - started >= maxWaitMs) {
      throw new PollRenderTimeout(renderId, maxWaitMs);
    }

    await sleep(interval, signal);
    interval = Math.min(maxIntervalMs, Math.round(interval * backoffFactor));
  }
}

function isTerminal(status: string): boolean {
  return (TERMINAL_STATUSES as readonly string[]).includes(status);
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    const t = setTimeout(() => {
      signal?.removeEventListener('abort', onAbort);
      resolve();
    }, ms);

    const onAbort = () => {
      clearTimeout(t);
      reject(new DOMException('Aborted', 'AbortError'));
    };

    signal?.addEventListener('abort', onAbort, { once: true });
  });
}

export class PollRenderTimeout extends Error {
  constructor(public readonly renderId: string, public readonly maxWaitMs: number) {
    super(`pollRender(${renderId}) timed out after ${maxWaitMs}ms`);
    this.name = 'PollRenderTimeout';
  }
}

export class PollRenderAborted extends Error {
  constructor(public readonly renderId: string) {
    super(`pollRender(${renderId}) aborted`);
    this.name = 'PollRenderAborted';
  }
}

File: sdks/python/philiprehberger_docgen_sdk/poll_render.py
"""
Poll a Docgen render until it reaches a terminal state.

Backs off exponentially (500ms → ~5s ceiling), respects a wall-clock
budget, raises PollRenderTimeout if the deadline elapses.
"""

from __future__ import annotations

import time
from typing import Optional

from philiprehberger_docgen.api.renders_api import RendersApi
from philiprehberger_docgen.models.render import Render


_TERMINAL_STATUSES = ("succeeded", "failed", "cancelled")


def poll_render(
    api: RendersApi,
    render_id: str,
    *,
    max_wait_ms: int = 60_000,
    initial_interval_ms: int = 500,
    max_interval_ms: int = 5_000,
    backoff_factor: float = 1.6,
) -> Render:
    """Poll until terminal. Raises PollRenderTimeout on overall timeout."""

    deadline = time.monotonic() + (max_wait_ms / 1000)
    interval_ms = initial_interval_ms

    while True:
        render = api.get_render(render_id=render_id)

        if render.status in _TERMINAL_STATUSES:
            return render

        if time.monotonic() >= deadline:
            raise PollRenderTimeout(render_id, max_wait_ms)

        time.sleep(interval_ms / 1000)
        interval_ms = min(max_interval_ms, int(interval_ms * backoff_factor))


class PollRenderTimeout(TimeoutError):
    """The render didn't reach a terminal state within max_wait_ms."""

    def __init__(self, render_id: str, max_wait_ms: int) -> None:
        super().__init__(f"poll_render({render_id}) timed out after {max_wait_ms}ms")
        self.render_id = render_id
        self.max_wait_ms = max_wait_ms
